### PR TITLE
Extend `ThatArePublicOrInternal` to also look at the setter of properties

### PR DESCRIPTION
**Pull request by Ruijin92**
## IMPORTANT 

* [x] If the PR touches the public API, the changes have been approved in a separate issue with the "api-approved" label.
* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).

### CODE REVIEW

File: Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs [hunk 1]
```diff
@@ -355,6 +355,32 @@ public void When_selecting_properties_return_types_it_should_return_the_correct_
...
+
+        // Act
+        IEnumerable<PropertyInfo> properties = type.Properties().ThatArePublicOrInternal.ToArray();
+
+        // Assert
+        properties.Should().HaveCount(3);
+    }
+
+    [Fact]
+    public void When_selecting_propeties_with_different_accessors_from_a_abstract_class_should_return_the_applicable_properties()
```
Comment (IT-VBFK) [inline Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs:373]:
```suggestion
    public void When_selecting_propeties_with_different_accessors_from_an_abstract_class_should_return_the_applicable_properties()
```

Comment (IT-VBFK) [reply Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs:373]: little typo

File: Src/FluentAssertions/Types/PropertyInfoSelector.cs [hunk 2]
```diff
@@ -40,16 +40,16 @@ public PropertyInfoSelector(IEnumerable<Type> types)
...
     {
         get
         {
             selectedProperties = selectedProperties.Where(property =>
             {
-                MethodInfo getter = property.GetGetMethod(nonPublic: true);
-                return (getter is not null) && (getter.IsPublic || getter.IsAssembly);
+                MethodInfo methodInfo = property.GetGetMethod(nonPublic: true) ?? property.GetSetMethod(nonPublic: true);
+                return (methodInfo is not null) && (methodInfo.IsPublic || methodInfo.IsAssembly);
             });
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/PropertyInfoSelector.cs:53]: The summary now says that _both_ getter and setter must be public/internal, but the implementation only requires one of them.

What about if both a getter and setter exists, but only one is public/internal?
Then there is an asymmetry.

Example
```cs
public class C
{
    public bool PrivateGet { private get; set; } // not included
    public bool PrivateSet { get; private set; } // included
}
```

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AJRgDMAbGMDASwh4A3NWoA3AIZR6AcxgZ6AXnoYAngAcYEDgAoAwgEpmAcQUAFKBC1R1ugESX+UjDDMZ7hgISiaZAJy65qy68hiGhr5MgQBSQpIAzqEKEb4S0vQJCsqqmtp6RqYWVjZ2jlDOkq4AygqePtTRQSFZ4ZGNAbpxeIm6raliVMAQEFz0wUElMLZqAJJ4HBD0GlMzhtQA3tT0O/RSMgC2CgAWEAAm84s5K9bT6kUY7gCyJ+e6eELmOMBc/GAgqlwMEM9AA/KDlqt7u5ahgXhhTmd3p9vr9/oCcMDfLtGAB2ei6I4I86XJb8BL0D6KPA4LhcEEAMgZBKJiNJzFmCS+Pz+9AIBHorJJCwgHISAEEElkDj81O0qABfQbDUb0bq9Sy3GakyFa9TrKhbKg4oj4m6lNQPZ6vJEfPDctEAjBAkHk+gbeich1/J1AhX0aA7D2cyXS2W+zFK4043b83UWh6w+GI5H21E+jHA+hu4Nc9Po52RgMyd2eiVSmAyrhqCMwBW+KODIgAZkYJHo+k2212LfoKrGThcbmyHpWlVccgUwkyU/oUZNrf74wqQ9hpbC07HQ5nGGnUYVQA=)

```cs
return property.GetGetMethod(nonPublic: true) is { IsPublic: true} or { IsAssembly: true}
    || property.GetSetMethod(nonPublic: true) is { IsPublic: true} or { IsAssembly: true};
```

Comment (Ruijin92) [reply Src/FluentAssertions/Types/PropertyInfoSelector.cs:53]: Yes, you're right!
 
 I will change the implemenation based on your suggestion.
 
 Summary: **Only select the properties that have atleast one public or internal accessor**
 
 Any other suggestion for the summary?

File: docs/_pages/releases.md [hunk 3]
```diff
@@ -20,6 +20,7 @@ sidebar:
 * Added new extension methods to be able to write `Exactly.Times(n)`, `AtLeast.Times(n)` and `AtMost.Times(n)` in a more fluent way - [#2047](https://github.com/fluentassertions/fluentassertions/pull/2047)
 
 ### Fixes
+* Fixed `ThatArePublicOrInternal` property of `PropertyInfoSelector` not including properties with set accessors - [#2084] (https://github.com/fluentassertions/fluentassertions/pull/2082)
```
Comment (jnyrup) [inline docs/_pages/releases.md:23]:
```suggestion
* `PropertyInfoSelector.ThatArePublicOrInternal` now takes the setter into account when determining if a property is `public` or `internal` - [#2082] (https://github.com/fluentassertions/fluentassertions/pull/2082)
```

File: Src/FluentAssertions/Types/PropertyInfoSelector.cs [hunk 4]
```diff
@@ -40,16 +40,16 @@ public PropertyInfoSelector(IEnumerable<Type> types)
     }
 
     /// <summary>
-    /// Only select the properties that have a public or internal getter.
+    /// Only select the properties that have atleast one public or internal accessor
```
Comment (jnyrup) [inline Src/FluentAssertions/Types/PropertyInfoSelector.cs:43]:
```suggestion
    /// Only select the properties that have at least one public or internal accessor
```

**Review by jnyrup — APPROVED** [review]

File: Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs [hunk 5]
```diff
@@ -355,6 +355,45 @@ public void When_selecting_properties_return_types_it_should_return_the_correct_
                 , typeof(string), typeof(int), typeof(int), typeof(int), typeof(int)
             });
     }
+
+    [Fact]
+    public void When_selecting_properties_with_only_set_accessors_it_should_return_the_applicable_properties()
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs:360]:
```suggestion
    public void When_a_property_only_has_a_public_setter_it_should_be_included_in_the_applicable_properties()
```

File: Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs [hunk 6]
```diff
@@ -355,6 +355,45 @@ public void When_selecting_properties_return_types_it_should_return_the_correct_
...
+    {
+        // Arrange
+        Type type = typeof(TestClassForPropertyInfoSelector);
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs:363]: 🔧 Although you just copied the way we did things before, I'd like us to start using an inline private class that only contains the relevant property(ies), right under the test. This makes the test much more readable and avoids unnecessarily long names like `TestClassForPropertyInfoSelector`. 

Same for the other tests you introduced.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]
Thank you for your contribution.

There's mostly remarks related to new insights I got during reviewing. Hope you're willing to address those.

It seems that this class can really benefit from some of our [recent design changes](https://github.com/fluentassertions/fluentassertions/blob/develop/Tests/FluentAssertions.Specs/Numeric/NullableNumericAssertionSpecs.cs#L193) such as grouping tests by their API and using the shorter naming conventions, but that's outside the scope of this PR.

File: Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs (hunk 5 above)
Comment (Ruijin92) [reply Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs:360]: So basically refactoring the `ThatArePublicOrInternal` tests under a public class `ThatArePublicOrInternal` i assume.

**Comment by Ruijin92** [comment]
So basically refactoring the `ThatArePublicOrInternal` tests under a public class `ThatArePublicOrInternal` i assume.

**Comment by dennisdoomen** [comment]
> So basically refactoring the ThatArePublicOrInternal tests under a public class ThatArePublicOrInternal i assume.

Yes. And it will make it possible to keep the names much shorter.

File: Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs (hunk 6 above)
Comment (Ruijin92) [reply Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs:363]: If i unterstand it correctly, you want the tests inside the newly created `ThatArePublicOrInternal` class and under each of the tests?

I saw on other tests, that you introduced the "HelperTestClasses" in a seperated file partialling it

Comment (dennisdoomen) [reply Tests/FluentAssertions.Specs/Types/PropertyInfoSelectorSpecs.cs:363]: > If i unterstand it correctly, you want the tests inside the newly created ThatArePublicOrInternal class and under each of the tests?

Yes. If possible, we want our tests to be as self-explanatory as possible. And using reusable test classes spreaded around the code base doesn't help understanding why the number of properties is expected to be "3"

> I saw on other tests, that you introduced the "HelperTestClasses" in a seperated file partialling it

That was a mistake ;-)

**Comment by Ruijin92** [comment]
- Avoid the use of **When** and **Should** in test names and use concise names like `Exclusion_of_missing_members_works_with_mapping`

Should this be also taking into account? As stated here 
[https://github.com/fluentassertions/fluentassertions/issues/1340](url)

**Comment by dennisdoomen** [comment]
> Should this be also taken into account? As stated here

Up to you

**Comment by Ruijin92** [comment]
Well, i leave it as it is right now.

**Review by dennisdoomen — APPROVED** [review]

**Pull request merged**